File: Src2/Icm.Core/Collections/SortedCollection/RangeIntervalEnumerator.vb
Namespace Icm.Collections.Generic

    ''' <summary>
    ''' Transforms an IEnumerator(Of T) into an IEnumerator(Of Vector2(Of T)) which enumerates
    ''' the consecutive pairs of the original IEnumerator.
    ''' </summary>
    ''' <remarks>
    ''' <para>For example, if the original IEnumerator contains three elements a, b, c, this enumerator
    ''' will have the following two elements: (a, b), (b, c).</para>
    ''' <para>If the original IEnumerator contains just one element x, this class will return a single
    ''' pair (x, x).</para>
    ''' <para>This enumerator will NOT dispose the original enumerator.</para>
    ''' </remarks>
    Public Class PairEnumerator(Of T)
        Implements IEnumerator(Of Vector2(Of T)), IEnumerable(Of Vector2(Of T))

        Private current_ As Vector2(Of T)
        Private ReadOnly pointIter_ As IEnumerator(Of T)

        Public Sub New(ByVal _pointIter As IEnumerator(Of T))
            pointIter_ = _pointIter
        End Sub

        Public ReadOnly Property Current() As Vector2(Of T) Implements IEnumerator(Of Vector2(Of T)).Current
            Get
                If current_ Is Nothing Then
                    Throw New InvalidOperationException("Enumerator has not been reset")
                End If
                Return current_
            End Get
        End Property

        Public ReadOnly Property Current1() As Object Implements System.Collections.IEnumerator.Current
            Get
                If current_ Is Nothing Then
                    Throw New InvalidOperationException("Enumerator has not been reset")
                End If
                Return current_
            End Get
        End Property

        Public Function MoveNext() As Boolean Implements System.Collections.IEnumerator.MoveNext
            Dim item2 As T

            If current_ Is Nothing Then
                pointIter_.MoveNext()
                Dim item1 = pointIter_.Current
                If pointIter_.MoveNext Then
                    item2 = pointIter_.Current
                Else
                    item2 = item1
                End If
                current_ = New Vector2(Of T)(item1, item2)
            Else
                If pointIter_.MoveNext Then
                    item2 = pointIter_.Current
                    current_ = New Vector2(Of T)(current_.Item2, item2)
                Else
                    current_ = Nothing
                    Return False
                End If
            End If
            Return True
        End Function


        Public Sub Reset() Implements IEnumerator.Reset
            current_ = Nothing
        End Sub

        Private disposedValue As Boolean = False        ' To detect redundant calls

        ' IDisposable
        Protected Overridable Sub Dispose(ByVal disposing As Boolean)
            If Not Me.disposedValue Then
                If disposing Then
                    ' free other state (managed objects).
                End If

                ' free your own state (unmanaged objects).
                ' set large fields to null.
            End If
            Me.disposedValue = True
        End Sub

#Region " IDisposable Support "
        ' This code added by Visual Basic to correctly implement the disposable pattern.
        Public Sub Dispose() Implements IDisposable.Dispose
            ' Do not change this code.  Put cleanup code in Dispose(ByVal disposing As Boolean) above.
            Dispose(True)
            GC.SuppressFinalize(Me)
        End Sub
#End Region

        Public Function GetEnumerator() As IEnumerator(Of Vector2(Of T)) Implements IEnumerable(Of Vector2(Of T)).GetEnumerator
            Return Me
        End Function

        Public Function GetEnumerator1() As IEnumerator Implements System.Collections.IEnumerable.GetEnumerator
            Return Me
        End Function
    End Class
End Namespace

File: Src/Icm.Core/Collections/SortedCollection/RangeIntervalEnumerator.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace Icm.Collections.Generic
{

	/// <summary>
	/// Transforms an IEnumerator(Of T) into an IEnumerator(Of Vector2(Of T)) which enumerates
	/// the consecutive pairs of the original IEnumerator.
	/// </summary>
	/// <remarks>
	/// <para>For example, if the original IEnumerator contains three elements a, b, c, this enumerator
	/// will have the following two elements: (a, b), (b, c).</para>
	/// <para>If the original IEnumerator contains just one element x, this class will return a single
	/// pair (x, x).</para>
	/// <para>This enumerator will NOT dispose the original enumerator.</para>
	/// </remarks>
	public class PairEnumerator<T> : IEnumerator<Vector2<T>>, IEnumerable<Vector2<T>>
	{

		private Vector2<T> current_;

		private readonly IEnumerator<T> pointIter_;
		public PairEnumerator(IEnumerator<T> _pointIter)
		{
			pointIter_ = _pointIter;
		}

		public Vector2<T> Current {
			get {
				if (current_ == null) {
					throw new InvalidOperationException("Enumerator has not been reset");
				}
				return current_;
			}
		}

		public object Current1 {
			get {
				if (current_ == null) {
					throw new InvalidOperationException("Enumerator has not been reset");
				}
				return current_;
			}
		}
		object System.Collections.IEnumerator.Current {
			get { return Current1; }
		}

		public bool MoveNext()
		{
			T item2 = default(T);

			if (current_ == null) {
				pointIter_.MoveNext();
				dynamic item1 = pointIter_.Current;
				if (pointIter_.MoveNext) {
					item2 = pointIter_.Current;
				} else {
					item2 = item1;
				}
				current_ = new Vector2<T>(item1, item2);
			} else {
				if (pointIter_.MoveNext) {
					item2 = pointIter_.Current;
					current_ = new Vector2<T>(current_.Item2, item2);
				} else {
					current_ = null;
					return false;
				}
			}
			return true;
		}


		public void Reset()
		{
			current_ = null;
		}

			// To detect redundant calls
		private bool disposedValue = false;

		// IDisposable
		protected virtual void Dispose(bool disposing)
		{
			if (!this.disposedValue) {
				if (disposing) {
					// free other state (managed objects).
				}

				// free your own state (unmanaged objects).
				// set large fields to null.
			}
			this.disposedValue = true;
		}

		#region " IDisposable Support "
		// This code added by Visual Basic to correctly implement the disposable pattern.
		public void Dispose()
		{
			// Do not change this code.  Put cleanup code in Dispose(ByVal disposing As Boolean) above.
			Dispose(true);
			GC.SuppressFinalize(this);
		}
		#endregion

		public IEnumerator<Vector2<T>> GetEnumerator()
		{
			return this;
		}

		public IEnumerator GetEnumerator1()
		{
			return this;
		}
		IEnumerator System.Collections.IEnumerable.GetEnumerator()
		{
			return GetEnumerator1();
		}
	}
}

//=======================================================
//Service provided by Telerik (www.telerik.com)
//Conversion powered by NRefactory.
//Twitter: @telerik
//Facebook: facebook.com/telerik
//=======================================================
